## CI triage — the 2 failing e2e lanes are pre-existing flakes on `main`, not regressions from this PR

**TL;DR:** Both red e2e checks fail with the *same* tests that are red on `main` right now (with zero Script-cache code present), and the failing app even rotates run-to-run. Neither failure mode is reachable by the change in this PR. `E2E Required Check` is the aggregation gate and is red only because those two lanes are red. The `@workflow/core` unit + replay-determinism suite is fully green on this branch.

### The two failures

1. **E2E Vercel Prod Tests (express)** — `AbortController > abortFromStepWorkflow: step abort cancels an in-flight sibling step`
   `AssertionError: expected false to be true` at `e2e.test.ts:3338` (`returnValue.stepSawAborted`). 1 failed / rest passed.

2. **E2E Local Prod Tests (nextjs-turbopack — stable lazyDiscovery enabled)** — `wellKnownAgentWorkflow (.well-known/agent)`
   `WorkflowNotRegisteredError: Workflow "…wellKnownAgentWorkflow" is not registered in the current deployment … build/bundling issue`. 1 failed / 151 passed.

### Evidence it's a pre-existing flake on `main`

Latest `main` push run [`2acf13cc7` / 27704378960](https://github.com/vercel/workflow/actions/runs/27704378960) — which contains **none** of this PR's changes — fails the same way:

| `main` lane (no Script-cache code) | Failing test | Assertion |
| --- | --- | --- |
| E2E Vercel Prod (tanstack-start) | `AbortController > abortFromStepWorkflow: step abort cancels an in-flight sibling step` | `expected false to be true` — **byte-identical to this PR's express failure** |
| E2E Vercel Prod (nextjs-turbopack) | `distributedAbortController - reconnect to existing controller` | same AbortController family |

And this PR's merge-base `main@cb181392b` ([run 27657696161](https://github.com/vercel/workflow/actions/runs/27657696161)) was already red on `E2E Vercel Prod (fastify)`. The failing app rotates between runs (fastify → tanstack-start → nextjs-turbopack → express) — the signature of a shared infra/timing flake in the deployed-Vercel e2e lane, not a code defect. The `wellKnownAgentWorkflow` / `…NotRegistered` family is the same non-deterministic build/bundling-registration flake that recurs across apps on `main`.

### Why the Script cache cannot be the cause

This PR's diff is 4 isolated files (`packages/core/src/vm/script-cache.ts` + its test, a ~23-line edit in `packages/core/src/workflow.ts`, and the changeset). It touches **only** the inline-replay VM-realm setup. It does not touch the workflow/step registration or manifest-lookup path, AbortController, or `@workflow/world-vercel`.

- The cache is keyed by `(workflowCode, filename)` and only ever returns a `vm.Script` compiled from *exactly that code string* — `vm.Script` compilation is pure, so a cache hit is byte-identical to a fresh `new Script(code, { filename })`. A miss-keyed cache would surface as **ReplayDivergenceError / wrong return values**, never as "workflow not registered" or AbortController signal timing.
- `WorkflowNotRegisteredError` is thrown at manifest lookup — **before any VM execution** — so it is upstream of the cache entirely.
- The AbortController assertion is about cross-network abort-signal propagation timing on a deployed app; the inline Script cache isn't in that path.

### Local verification on this branch (`pgp/perf-reusable-vm-replay`, commit `ef7655dc5`)

- `pnpm --filter @workflow/core build` — clean
- `cd packages/core && pnpm test` — **55 files / 1241 tests passed**, incl. `workflow.test.ts` (79 replay/determinism tests), `abort-controller-step.test.ts` (18), and `vm/script-cache.test.ts` (5)
- `vm/script-cache.test.ts` run explicitly — 5/5 pass

No code change is warranted for these CI failures; they are environmental flakes shared with `main`.

--- a/packages/core/src/vm/script-cache.test.ts
+++ b/packages/core/src/vm/script-cache.test.ts
@@ -5,6 +5,7 @@ import {
   clearWorkflowScriptCache,
   getCachedWorkflowScript,
   runCachedWorkflowScript,
+  workflowScriptCacheSize,
 } from './script-cache.js';
 
 const seed = 'script-cache seed';
@@ -16,6 +17,25 @@ globalThis.__private_workflows.set('my/workflow', async function workflow(name) 
   return 'hello,' + name + ',' + Math.random() + ',' + Date.now();
 });
 `;
+
+/**
+ * Builds a realistic multi-workflow bundle: many registered workflow functions
+ * (the production shape) with a distinguishing `marker` baked in so two bundles
+ * built with different markers are genuinely different code (not the trivial
+ * `1 + 1` / `2 + 2` strings). Each workflow returns a value derived from the
+ * marker so a mis-served Script would produce a detectably wrong result.
+ */
+function buildBundle(marker: string, workflowCount = 12): string {
+  const defs: string[] = [];
+  for (let i = 0; i < workflowCount; i++) {
+    defs.push(
+      `globalThis.__private_workflows.set('app/workflow-${i}', async function workflow${i}(name) {\n` +
+        `  return '${marker}:' + ${i} + ':' + name + ':' + Math.random();\n` +
+        `});`
+    );
+  }
+  return `globalThis.__private_workflows = new Map();\n${defs.join('\n')}\n`;
+}
 
 describe('script-cache', () => {
   afterEach(() => {
@@ -90,5 +110,93 @@ describe('script-cache', () => {
     // Deterministic context => identical results every run.
     expect(results[0]).toEqual(results[1]);
     expect(results[1]).toEqual(results[2]);
+  });
+
+  it('bounds the bundle cache and evicts least-recently-used bundles', () => {
+    // Insert far more distinct bundles than any sane cap, simulating a long
+    // dev/watch session where every edit produces a new bundle string. The
+    // cache must NOT grow monotonically with edit count.
+    const editCount = 100;
+    const filename = 'workflows/a.ts';
+    for (let i = 0; i < editCount; i++) {
+      getCachedWorkflowScript(buildBundle(`edit-${i}`), filename);
+    }
+
+    const size = workflowScriptCacheSize();
+    expect(size).toBeGreaterThan(0);
+    // Bounded well below the number of edits — the whole point of the LRU.
+    expect(size).toBeLessThan(editCount);
+
+    // The cache still serves correctly after heavy churn: the most-recently
+    // inserted bundle is retained and repeated lookups return the same Script.
+    const latest = buildBundle(`edit-${editCount - 1}`);
+    expect(getCachedWorkflowScript(latest, filename)).toBe(
+      getCachedWorkflowScript(latest, filename)
+    );
+  });
+
+  it('keeps the most-recently-used bundle and evicts the stale one', () => {
+    const filename = 'workflows/a.ts';
+    // Seed an "old" bundle, then keep it hot by re-touching it while many
+    // unrelated bundles churn through. LRU must NOT evict the bundle we keep
+    // using, even though it was inserted first.
+    const hot = buildBundle('hot');
+    const hotScript = getCachedWorkflowScript(hot, filename);
+
+    for (let i = 0; i < 50; i++) {
+      getCachedWorkflowScript(buildBundle(`cold-${i}`), filename);
+      // Re-access the hot bundle so it stays most-recently-used.
+      expect(getCachedWorkflowScript(hot, filename)).toBe(hotScript);
+    }
+
+    // After all that churn the hot bundle is still the *same* cached Script —
+    // proving LRU recency (touch-on-access), not mere insertion order, governs
+    // eviction.
+    expect(getCachedWorkflowScript(hot, filename)).toBe(hotScript);
+  });
+
+  it('never returns the wrong Script across realistic multi-workflow bundles', async () => {
+    // Two genuinely different bundles (production-shape: many workflows each),
+    // distinguished by their marker, plus two filenames. Every distinct
+    // (code, filename) must map to its own Script, and running each must
+    // produce results derived from its own marker — never another bundle's.
+    const bundleX = buildBundle('bundle-X');
+    const bundleY = buildBundle('bundle-Y');
+    const fileA = 'workflows/a.ts';
+    const fileB = 'workflows/b.ts';
+
+    const xa = getCachedWorkflowScript(bundleX, fileA);
+    const xb = getCachedWorkflowScript(bundleX, fileB);
+    const ya = getCachedWorkflowScript(bundleY, fileA);
+    const yb = getCachedWorkflowScript(bundleY, fileB);
+
+    // All four (code, filename) combinations are distinct Script objects.
+    const scripts = [xa, xb, ya, yb];
+    for (let i = 0; i < scripts.length; i++) {
+      for (let j = i + 1; j < scripts.length; j++) {
+        expect(scripts[i]).not.toBe(scripts[j]);
+      }
+    }
+
+    // Same (code, filename) is stable across lookups.
+    expect(getCachedWorkflowScript(bundleX, fileA)).toBe(xa);
+    expect(getCachedWorkflowScript(bundleY, fileB)).toBe(yb);
+
+    // Running each bundle yields its OWN marker, confirming no cross-wiring.
+    const { context: ctxX } = createContext({ seed, fixedTimestamp });
+    runCachedWorkflowScript(bundleX, fileA, ctxX);
+    const fnX = runInContext(
+      `globalThis.__private_workflows?.get('app/workflow-3')`,
+      ctxX
+    ) as (n: string) => Promise<string>;
+    expect(await fnX('z')).toContain('bundle-X:3:z');
+
+    const { context: ctxY } = createContext({ seed, fixedTimestamp });
+    runCachedWorkflowScript(bundleY, fileA, ctxY);
+    const fnY = runInContext(
+      `globalThis.__private_workflows?.get('app/workflow-3')`,
+      ctxY
+    ) as (n: string) => Promise<string>;
+    expect(await fnY('z')).toContain('bundle-Y:3:z');
   });
 });

--- a/packages/core/src/vm/script-cache.ts
+++ b/packages/core/src/vm/script-cache.ts
@@ -19,24 +19,74 @@ import { type Context, Script } from 'node:vm';
  * no realm/context state — it is only bound to a context at `runInContext`
  * time. So a single compiled `Script` can be reused across replays AND across
  * workflow invocations in the same process without affecting determinism: the
- * produced workflow function and any thrown errors are byte-identical to the
- * previous re-parse-every-time behaviour.
+ * produced workflow function is identical to the previous re-parse-every-time
+ * behaviour, with identical `filename` source attribution (see the precise
+ * claim — and its one caveat — in `runWorkflow`).
  *
  * Keying
  * ------
- * Keyed by `code` then `filename`. The `filename` is part of the key because
- * it is baked into the compiled script's source attribution and surfaces in
- * stack traces (and is consumed by `remapErrorStack` at runtime). Two
- * workflows in the same bundle share the same `code` but can have different
- * `filename`s, so they must not share a compiled `Script`. The number of
- * distinct `(code, filename)` pairs in a process is bounded by
- * (bundle versions) × (workflow names) and is small in practice.
+ * Keyed by `code` then `filename`. The `filename` is part of the key on
+ * purpose, NOT as a dedupe key: it is baked into the compiled script's source
+ * attribution and surfaces in stack traces, where `remapErrorStack` keys on it
+ * to map frames back to the user's source. Two workflows in the same bundle
+ * share the same `code` but have different `filename`s, so they intentionally
+ * compile to distinct `Script`s — collapsing them onto a single shared `Script`
+ * would misattribute one workflow's stack frames to another file. The cost of
+ * keeping them distinct is that the whole bundle is compiled once per distinct
+ * `filename` (not once per bundle); in practice that is bounded by the number
+ * of source files that define a workflow, and because V8 lazily compiles
+ * function bodies the duplicated work is the (cheap) top-level parse, not full
+ * per-workflow codegen.
  *
- * We use a nested Map (code -> filename -> Script) so that swapping the bundle
- * (e.g. a new deployment/hot-reload producing a different `code`) lets the old
- * code string and all its per-filename scripts become unreachable together.
+ * We use a nested Map (code -> filename -> Script) so that evicting a bundle
+ * (e.g. a new deployment/hot-reload producing a different `code`) drops the old
+ * code string and all of its per-filename scripts together.
+ *
+ * Bounding
+ * --------
+ * The top-level (`code`-keyed) map is an insertion-ordered LRU capped at
+ * `MAX_BUNDLES` entries. In production this bound is never reached: a
+ * deployment is its own process serving exactly one build-time bundle literal
+ * (skew protection runs old versions as separate processes), so there is a
+ * single `code` key for the process lifetime. The bound exists for dev/watch
+ * mode, where the dev route re-reads `workflowCode` from disk and re-invokes
+ * the entrypoint on every edit — each edit produces a NEW bundle string, which
+ * without a bound would pin every historical version forever (~0.8MB per edit,
+ * growing monotonically with edit count). The dev path only ever needs the
+ * latest bundle, so an LRU that keeps the few most-recent bundles and evicts
+ * the rest preserves the pre-cache GC behaviour while still serving the
+ * steady-state single-bundle case for free. The per-`filename` inner map is not
+ * separately bounded: it is naturally bounded by the (small) number of workflow
+ * source files in a bundle and is dropped wholesale when its parent `code`
+ * entry is evicted.
  */
 const scriptCache = new Map<string, Map<string, Script>>();
+
+/**
+ * Max number of distinct bundle (`code`) versions to retain. One is enough for
+ * production; a handful covers pathological dev hot-reload / repeated-rebuild
+ * churn within a single long-lived process (e.g. a watch session or a test
+ * file) without unbounded growth. Kept deliberately small — there is no value
+ * in retaining stale bundles, only a memory cost.
+ */
+const MAX_BUNDLES = 8;
+
+/**
+ * Looks up the per-filename map for `code`, marking it most-recently-used.
+ * Relies on `Map` preserving insertion order: deleting and re-inserting an
+ * existing key moves it to the end (newest), so the first key is always the
+ * least-recently-used eviction candidate.
+ */
+function touchBundle(code: string): Map<string, Script> | undefined {
+  const byFilename = scriptCache.get(code);
+  if (byFilename === undefined) {
+    return undefined;
+  }
+  // Move to the most-recently-used position (end of insertion order).
+  scriptCache.delete(code);
+  scriptCache.set(code, byFilename);
+  return byFilename;
+}
 
 /**
  * Returns a compiled `vm.Script` for the given workflow bundle code and
@@ -52,10 +102,19 @@ export function getCachedWorkflowScript(
   code: string,
   filename: string
 ): Script {
-  let byFilename = scriptCache.get(code);
+  let byFilename = touchBundle(code);
   if (byFilename === undefined) {
     byFilename = new Map<string, Script>();
     scriptCache.set(code, byFilename);
+    // Evict the least-recently-used bundle(s) when over the cap. New bundles
+    // are appended at the end, so the oldest live at the front.
+    while (scriptCache.size > MAX_BUNDLES) {
+      const oldest = scriptCache.keys().next().value;
+      if (oldest === undefined) {
+        break;
+      }
+      scriptCache.delete(oldest);
+    }
   }
   let script = byFilename.get(filename);
   if (script === undefined) {
@@ -83,4 +142,12 @@ export function runCachedWorkflowScript(
  */
 export function clearWorkflowScriptCache(): void {
   scriptCache.clear();
+}
+
+/**
+ * Number of distinct bundle (`code`) versions currently retained. Exposed for
+ * tests asserting the LRU bound; not used on the hot path.
+ */
+export function workflowScriptCacheSize(): number {
+  return scriptCache.size;
 }

--- a/packages/core/src/workflow.ts
+++ b/packages/core/src/workflow.ts
@@ -42,11 +42,11 @@ import {
   createAbortSignalStatics,
   createCreateAbortController,
 } from './workflow/abort-controller.js';
+import { createSetAttributes } from './workflow/attribute-dispatcher.js';
 import type { WorkflowMetadata } from './workflow/get-workflow-metadata.js';
 import { WORKFLOW_CONTEXT_SYMBOL } from './workflow/get-workflow-metadata.js';
 import { createCreateHook } from './workflow/hook.js';
 import { createSleep } from './workflow/sleep.js';
-import { createSetAttributes } from './workflow/attribute-dispatcher.js';
 
 /**
  * Drain pending queue items at workflow completion (success or failure).
@@ -773,14 +773,20 @@ export async function runWorkflow(
     // process-wide cache of the compiled `vm.Script`. The bundle is the same
     // string for every replay and every invocation in this process, and
     // compilation is a pure function of `(code, filename)`, so reusing the
-    // compiled Script across replays is determinism-safe: it produces a
-    // byte-identical result to re-parsing the bundle every time, but skips the
-    // (expensive) re-parse. Evaluating the bundle registers every workflow on
+    // compiled Script across replays is determinism-safe: it produces the same
+    // workflow function and the same `filename` source attribution as
+    // re-parsing the bundle every time, but skips the (expensive) re-parse.
+    // Evaluating the bundle registers every workflow on
     // `globalThis.__private_workflows`; the trailing lookup expression then
     // retrieves the requested workflow function. The lookup is evaluated as a
-    // separate cached Script under the same `filename` so its source
-    // attribution (and thus stack traces) match the previous combined-string
-    // behaviour.
+    // separate cached Script under the same `filename`, so error stack frames
+    // still attribute to the workflow's source file (`remapErrorStack` keys on
+    // `filename`). The one behavioural difference from the previous
+    // single-combined-string approach is the *line number* of an error thrown
+    // by the lookup expression itself: it now reports line 1 of the lookup
+    // Script rather than the line just past the end of the bundle. That path
+    // is rare (it requires the lookup `?.get(...)` expression to throw) and
+    // does not affect the workflow function or replay determinism.
     runCachedWorkflowScript(workflowCode, filename, context);
     const workflowFn = runCachedWorkflowScript(
       `globalThis.__private_workflows?.get(${JSON.stringify(workflowRun.workflowName)})`,

